### PR TITLE
expression: Add `GetUserVarsReader` to `EvalContext`

### LINKED ISSUE #55865: Add `GetUserVarsReader` to `EvalContext`

### PR DESCRIPTION
**Pull request by lcwangchao**
### What problem does this PR solve?

Issue Number: close #55865

### What changed and how does it work?

This PR adds new method `GetUserVarsReader` to `EvalContext` so that the user vars can be read directly from the context. It has some benefits:

1. `BuildSimpleExpr` can now build expressions like `@a+1` without adding optional property `SessionVars` to the inner context.
2. We can evaluate an expression like `@a+1` or `get_var("a")` without requiring an additional property `SessionVars` in `EvalContext`.
3. Lazy cursor fetch can now support SQL with user vars.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/sessionctx/variable/session.go [hunk 1]
```diff
@@ -2391,15 +2431,6 @@ func (s *SessionVars) SetStringUserVar(name string, strVal string, collation str
 	}
 }
 
-// UnsetUserVar unset an user defined variable by name.
-func (s *SessionVars) UnsetUserVar(varName string) {
```
Comment (lcwangchao) [inline pkg/sessionctx/variable/session.go:2395]: `UnsetUserVar` has been moved to: https://github.com/pingcap/tidb/pull/55866/files#diff-1df3757be8063f9a9a3c70650f58368402d93bbf842ac2e1c2f8426542f1a2f5R743

**Review by YangKeao — APPROVED** [review]
LGTM

**Review by XuHuaiyu — APPROVED** [review]

**Review by qw4990 — APPROVED** [review]

**Review by yudongusa — APPROVED** [review]
LGTM for the variable

**Pull request merged**